### PR TITLE
Created an endpoint of obtaining a specific offer

### PR DESCRIPTION
**Pull request by AndriiChepeliuk**
Created an endpoint of obtaining a specific offer

### CODE REVIEW

File: Core/DTO/OfferDTO/OfferInfoDTO.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+        public string Description { get; set; }
+        public bool IsClosed { get; set; }
+        public float GoodsWeight { get; set; }
+        public DateTimeOffset StartDate { get; set; }
+        public DateTimeOffset ExpirationDate { get; set; }
+        public string GoodCategory { get; set; }
+        public string Role { get; set; }
+        public int? RelatedTripId { get; set; }
+        public PointInfoDTO Point { get; set; }
+    }
```
Comment (YevheniiVovk) [inline Core/DTO/OfferDTO/OfferInfoDTO.cs]: Why ID is not presented ?

File: Core/DTO/PointDTO/PointInfoDTO.cs [hunk 2]
```diff
@@ -1,6 +1,6 @@
 ﻿namespace Core.DTO.PointDTO
 {
-    public class PointCreateDTO
+    public class PointInfoDTO
     {
         public float Latitude { get; set; }
         public float Longitude { get; set; }
```
Comment (YevheniiVovk) [inline Core/DTO/PointDTO/PointInfoDTO.cs]: How many point's will be in Kiev - Rivne route

File: Core/Services/OfferService.cs [hunk 3]
```diff
@@ -16,24 +22,33 @@ public class OfferService : IOfferService
...
+            IGoodCategoryService goodCategoryService,
+            IRoleService roleService,
+            IPointService pointService,
+            IRepository<Role> roleRepository,
+            IRepository<GoodCategory> goodCategoryRepository,
+            IRepository<Point> pointRepository)
         {
             _pointService = pointService;
-            _roleRepository = roleRepository;
+            _roleService = roleService;
```
Comment (YevheniiVovk) [inline Core/Services/OfferService.cs]: Please use already exist RoleManager class

File: Core/Services/OfferService.cs [hunk 4]
```diff
@@ -44,14 +59,39 @@ public async Task<OfferCreateDTO> CreateOfferAsync(OfferCreateDTO offerCreate, s
...
+        {
+            var offer = await _offerRepository.GetByIdAsync(offerId);
+
+            ExceptionMethods.OfferNullCheck(offer);
+            ExceptionMethods.UserNullCheck(await _userManager.FindByIdAsync(userId));
+
+            if (offer.OfferCreatorId != userId)
+            {
+                throw new HttpException(ErrorMessages.OfferNotFound, HttpStatusCode.NotFound);
+            }
```
Comment (YevheniiVovk) [inline Core/Services/OfferService.cs]: Are you sure that BL should know anything about web UI ?

File: Core/Services/OfferService.cs [hunk 5]
```diff
@@ -44,14 +59,39 @@ public async Task<OfferCreateDTO> CreateOfferAsync(OfferCreateDTO offerCreate, s
...
+
+            var offerInfo = _mapper.Map<OfferInfoDTO>(offer);
+            offerInfo.Role = (await _roleRepository.GetByIdAsync(offer.CreatorRoleId)).Name;
```
Comment (YevheniiVovk) [inline Core/Services/OfferService.cs]: Please use role manager

File: Core/Services/OfferService.cs [hunk 6]
```diff
@@ -44,14 +59,39 @@ public async Task<OfferCreateDTO> CreateOfferAsync(OfferCreateDTO offerCreate, s
...
+            offerInfo.GoodCategory = (await _goodCategoryRepository.
+                GetByIdAsync(offer.GoodCategoryId)).Name;
+            var point = await _pointRepository.GetByIdAsync(offer.OfferPointId);
+
+            ExceptionMethods.PointNullCheck(point);
```
Comment (YevheniiVovk) [inline Core/Services/OfferService.cs]: Are you sure that this is the good way of check?

File: Core/Services/OfferService.cs [hunk 7]
```diff
@@ -44,14 +59,39 @@ public async Task<OfferCreateDTO> CreateOfferAsync(OfferCreateDTO offerCreate, s
             offer.OfferCreatorId = userId;
             offer.CreationDate = DateTimeOffset.UtcNow;
             offer.IsClosed = false;
-            offer.CreatorRoleId = _roleRepository.GetRoleByName(offerCreate.Role);
-            offer.GoodCategoryId = _goodCategoryRepository
+            offer.CreatorRoleId = _roleService.GetRoleByName(offerCreate.Role);
```
Comment (YevheniiVovk) [inline Core/Services/OfferService.cs]: Use role manager.

File: Core/Services/OfferService.cs [hunk 8]
```diff
@@ -44,14 +59,39 @@ public async Task<OfferCreateDTO> CreateOfferAsync(OfferCreateDTO offerCreate, s
...
+            offer.GoodCategoryId = _goodCategoryService
                 .GetGoodCategoryByName(offerCreate.GoodCategory);
             offer.OfferPointId = await _pointService.CreatePointForOfferAsync(offerCreate.Point);
 
             await _offerRepository.InsertAsync(offer);
             await _offerRepository.SaveChangesAsync();
             return _mapper.Map<OfferCreateDTO>(offerCreate);
```
Comment (YevheniiVovk) [inline Core/Services/OfferService.cs:56]: Why we return a new one object?

File: Core/Helpers/ApplicationProfiles/OfferProfiles.cs [hunk 9]
```diff
@@ -12,6 +12,10 @@ public OfferProfiles()
                 .ForMember(offer => offer.Point, dto => dto.Ignore())
                 .ForMember(offer => offer.Role, dto => dto.Ignore())
                 .ForMember(offer => offer.GoodCategory, dto => dto.Ignore());
+            CreateMap<OfferInfoDTO, Offer>().ReverseMap()
+                .ForMember(dto => dto.Point, offer => offer.Ignore())
+                .ForMember(dto => dto.Role, offer => offer.Ignore())
+                .ForMember(dto => dto.GoodCategory, offer => offer.Ignore());
```
Comment (YevheniiVovk) [inline Core/Helpers/ApplicationProfiles/OfferProfiles.cs]: why offer is ignored ?

File: Core/DTO/OfferDTO/OfferCreateDTO.cs [hunk 10]
```diff
@@ -11,6 +11,6 @@ public class OfferCreateDTO
         public DateTimeOffset ExpirationDate { get; set; }
         public string GoodCategory { get; set; }
         public string Role { get; set; }
-        public PointCreateDTO Point { get; set; }
+        public PointInfoDTO Point { get; set; }
```
Comment (YevheniiVovk) [inline Core/DTO/OfferDTO/OfferCreateDTO.cs]: this is bad idea. details described down in db area

File: API/Controllers/OfferController.cs [hunk 11]
```diff
@@ -27,5 +27,15 @@ public async Task<IActionResult> Create([FromBody] OfferCreateDTO offer)
             var userId = _userService.GetCurrentUserNameIdentifier(User);
             return Ok(await _offerService.CreateOfferAsync(offer, userId));
         }
+
+        [HttpGet("offer-info")]
+        public async Task<IActionResult> GetOfferInfo(int offerId)
+        {
+            var userId = _userService.GetCurrentUserNameIdentifier(User);
+
+            OfferInfoDTO offerDTO = await _offerService.GetConcreteOfferAsync(offerId, userId);
```
Comment (marynakernychna) [inline API/Controllers/OfferController.cs]: You do not have to specify the type of data being returned. You can just write «var»

File: Core/DTO/PointDTO/PointInfoDTO.cs (hunk 2 above)
Comment (marynakernychna) [reply Core/DTO/PointDTO/PointInfoDTO.cs:6]: As much as the driver can / wants to cover

File: Core/Services/OfferService.cs (hunk 7 above)
Comment (marynakernychna) [reply Core/Services/OfferService.cs:62]: We have two tables for roles. This is not for Identity Role

File: API/Controllers/OfferController.cs (hunk 11 above)
Comment (AndriiChepeliuk) [reply API/Controllers/OfferController.cs]: Resolved

File: Core/DTO/OfferDTO/OfferCreateDTO.cs (hunk 10 above)
Comment (AndriiChepeliuk) [reply Core/DTO/OfferDTO/OfferCreateDTO.cs]: what do you mean?

File: Core/DTO/OfferDTO/OfferInfoDTO.cs (hunk 1 above)
Comment (AndriiChepeliuk) [reply Core/DTO/OfferDTO/OfferInfoDTO.cs]: Added

File: Core/Services/OfferService.cs (hunk 5 above)
Comment (marynakernychna) [reply Core/Services/OfferService.cs]: We have two tables for roles. This is not for Identity Role

File: Core/Services/OfferService.cs (hunk 6 above)
Comment (marynakernychna) [reply Core/Services/OfferService.cs]: Why not?

File: Core/Services/OfferService.cs (hunk 4 above)
Comment (marynakernychna) [reply Core/Services/OfferService.cs]: What if we get a request not from web UI, how do we check it?

File: Core/DTO/PointDTO/PointInfoDTO.cs (hunk 2 above)
Comment (marynakernychna) [reply Core/DTO/PointDTO/PointInfoDTO.cs]: As much as the driver can / wants to cover

File: Core/Services/OfferService.cs (hunk 7 above)
Comment (marynakernychna) [reply Core/Services/OfferService.cs]: We have two tables for roles. This is not for Identity Role

File: Core/Helpers/ApplicationProfiles/OfferProfiles.cs (hunk 9 above)
Comment (AndriiChepeliuk) [reply Core/Helpers/ApplicationProfiles/OfferProfiles.cs]: I agree, this is superfluous

File: Core/Services/OfferService.cs (hunk 3 above)
Comment (AndriiChepeliuk) [reply Core/Services/OfferService.cs]: here we do not use IdentityRole

File: Core/Services/OfferService.cs (hunk 8 above)
Comment (AndriiChepeliuk) [reply Core/Services/OfferService.cs:56]: Marina will solve this problem in her next task

File: Core/Services/OfferService.cs [hunk 12]
```diff
@@ -44,14 +58,33 @@ public async Task<OfferCreateDTO> CreateOfferAsync(OfferCreateDTO offerCreate, s
...
+
+        public async Task<OfferInfoDTO> GetOfferByIdAsync(int offerId, string userId)
+        {
+            var offer = _offerRepository.Query().FirstOrDefault(o => o.Id == offerId && o.OfferCreatorId == userId);
+
+            ExceptionMethods.OfferNullCheck(offer);
+
+            var offerInfo = _mapper.Map<OfferInfoDTO>(offer);
+            offerInfo.Role = (await _roleRepository.GetByIdAsync(offer.CreatorRoleId)).Name;
+            offerInfo.GoodCategory = (await _goodCategoryRepository.
```
Comment (Grasmus) [inline Core/Services/OfferService.cs]: I think we should check all entities from db for null

Comment (AndriiChepeliuk) [reply Core/Services/OfferService.cs]: done!

File: Core/Services/OfferService.cs [hunk 13]
```diff
@@ -21,33 +18,24 @@ public class OfferService : IOfferService
         private readonly IMapper _mapper;
         private readonly IRepository<Offer> _offerRepository;
         private readonly UserManager<User> _userManager;
-        private readonly IGoodCategoryService _goodCategoryService;
-        private readonly IRoleService _roleService;
+        private readonly IGoodCategoryService _goodCategoryRepository;
```
Comment (Grasmus) [inline Core/Services/OfferService.cs]: Why do we name service as repository?

File: Core/Services/OfferService.cs [hunk 14]
```diff
@@ -21,33 +18,24 @@ public class OfferService : IOfferService
...
+        private readonly IRoleService _roleRepository;
```
Comment (Grasmus) [inline Core/Services/OfferService.cs]: Same here

File: Core/Services/OfferService.cs [hunk 15]
```diff
@@ -53,5 +55,23 @@ public async Task<OfferCreateDTO> CreateOfferAsync(OfferCreateDTO offerCreate, s
...
+                .Where(o => o.Id == offerId && o.OfferCreatorId == userId)
+                .Include(offer => offer.Point)
+                .Include(offer => offer.Role)
+                .Include(offer => offer.GoodCategory)
+                .FirstOrDefaultAsync();
+
+            ExceptionMethods.OfferNullCheck(offer);
+
+            var offerInfo = _mapper.Map<OfferInfoDTO>(offer);
+            offerInfo.Role = offer.Role.Name;
```
Comment (Eugene-pas) [inline Core/Services/OfferService.cs]: AutoMapper can do this

**Review by Eugene-pas — APPROVED** [review]

**Review by Grasmus — APPROVED** [review]

**Pull request merged**